apm-agent-plugins/apm-javalin-plugin/src/main/java/co/elastic/apm/agent/javalin/JavalinInstrumentation.java

```diff
@@ -0,0 +1,167 @@
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ * #L%
+ */
+package co.elastic.apm.agent.javalin;
+
+import co.elastic.apm.agent.bci.TracerAwareInstrumentation;
+import co.elastic.apm.agent.impl.transaction.AbstractSpan;
+import co.elastic.apm.agent.impl.transaction.Span;
+import co.elastic.apm.agent.impl.transaction.Transaction;
+import co.elastic.apm.agent.util.VersionUtils;
+import io.javalin.http.Context;
+import io.javalin.http.Handler;
+import io.javalin.http.HandlerType;
+import net.bytebuddy.asm.Advice;
+import net.bytebuddy.description.method.MethodDescription;
+import net.bytebuddy.description.type.TypeDescription;
+import net.bytebuddy.matcher.ElementMatcher;
+
+import javax.annotation.Nullable;
+import java.lang.reflect.Field;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.concurrent.CompletableFuture;
+
+import static co.elastic.apm.agent.impl.transaction.AbstractSpan.PRIO_HIGH_LEVEL_FRAMEWORK;
+import static net.bytebuddy.matcher.ElementMatchers.hasSuperType;
+import static net.bytebuddy.matcher.ElementMatchers.isInterface;
+import static net.bytebuddy.matcher.ElementMatchers.named;
+import static net.bytebuddy.matcher.ElementMatchers.not;
+import static net.bytebuddy.matcher.ElementMatchers.takesArgument;
+
+public class JavalinInstrumentation extends TracerAwareInstrumentation {
+
+    private static final String FRAMEWORK_NAME = "Javalin";
+
+    @Override
+    public ElementMatcher<? super TypeDescription> getTypeMatcher() {
+        return hasSuperType(named("io.javalin.http.Handler")).and(not(isInterface()));
+    }
+
+    @Override
+    public ElementMatcher<? super MethodDescription> getMethodMatcher() {
+        return named("handle").and(takesArgument(0, named("io.javalin.http.Context")));
+    }
+
+    @Override
+    public Collection<String> getInstrumentationGroupNames() {
+        return Collections.singleton("javalin");
+    }
+
+    @Override
+    public String getAdviceClassName() {
+        return "co.elastic.apm.agent.javalin.JavalinInstrumentation$HandlerAdapterAdvice";
+    }
+
+    public static class HandlerAdapterAdvice {
+
+        private static Field handlerTypeField = null;
+
+        @Advice.OnMethodEnter(suppress = Throwable.class, inline = false)
+        public static Object setSpanAndTransactionName(@Advice.This Handler handler,
+                                                       @Advice.Argument(0) Object ctxObject) throws Exception {
+            final Transaction transaction = tracer.currentTransaction();
+            if (transaction == null) {
+                return null;
+            }
+            Context ctx = ((Context) ctxObject);
+
+            // TODO FIXME I don't think this is the right way, I am sure there is a fancy bytecode trick that I am not aware off
```

Comment:
Yes, there is! You can use `@Advice.FieldValue`. See [example](https://github.com/elastic/apm-agent-java/blob/master/apm-agent-plugins/apm-okhttp-plugin/src/main/java/co/elastic/apm/agent/okhttp/OkHttpClientInstrumentation.java#L58)

Reply:
I saw `Advice.FieldValue`, however this seems to be about the class that is being instrumented, which in this case would be the `Handler` functional interface. I need to access a field value within the argument of the instrumented class.

Am I wrong on that one and there is a way? :-)

Reply:
Ohh sorry, you are right, it is the `Context` field you are after. 
Let me see if I can suggest something

Reply:
OK, if I am correct and `addHandler` is what gets called every time a `Handler` is submitted to `handle`, then you can use a global **weakly referenced** mapping between a `Handler` **instance** and the corresponding type. So, when `addHandler` is called, you map the `HandlerType` (as the map value) to the `Handler` instance (as the map key) and when `handle` ends, you clear this entry from the map.
That is preferable over getting the field value through reflection. In general, reflection is problematic because it may affect the class loading order, something that Byte Buddy does a lot to prevent. Also, relying on API tends to be more stable than relying on internal stuff, like a field.

Reply:
Comment above is irrelevant as the `addHandler` is indeed only part of the setup flow, not part of the request handling flow.